lib/src/voip/group_call_session.dart

```diff
@@ -0,0 +1,1361 @@
+ *   published by the Free Software Foundation, either version 3 of the
+ *   License, or (at your option) any later version.
+ *
+ *   This program is distributed in the hope that it will be useful,
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *   GNU Affero General License for more details.
+ *
+ *   You should have received a copy of the GNU Affero General License
+ *   along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+import 'dart:async';
+import 'dart:convert';
+import 'dart:core';
+import 'dart:typed_data';
+
+import 'package:collection/collection.dart';
+import 'package:webrtc_interface/webrtc_interface.dart';
+
+import 'package:matrix/matrix.dart';
+import 'package:matrix/src/utils/cached_stream_controller.dart';
+import 'package:matrix/src/utils/crypto/crypto.dart';
+import 'package:matrix/src/voip/models/call_membership.dart';
+import 'package:matrix/src/voip/models/call_options.dart';
+import 'package:matrix/src/voip/models/voip_id.dart';
+import 'package:matrix/src/voip/utils/stream_helper.dart';
+
+/// Holds methods for managing a group call. This class is also responsible for
+/// holding and managing the individual `CallSession`s in a group call.
+class GroupCallSession {
+  // Config
+  static const updateExpireTsTimerDuration = Duration(seconds: 15);
+  static const expireTsBumpDuration = Duration(seconds: 45);
+  static const activeSpeakerInterval = Duration(seconds: 5);
+
+  final Client client;
+  final VoIP voip;
+  final Room room;
+
+  /// is a list of backend to allow passing multiple backends in the future
+  /// we use the first backend everywhere as of now
+  final List<CallBackend> backends;
+  final String? application;
+  final String? scope;
+
+  GroupCallState state = GroupCallState.localCallFeedUninitialized;
+  StreamSubscription<CallSession>? _callSubscription;
+
+  /// participant:volume
+  final Map<CallParticipant, double> audioLevelsMap = {};
+  CallParticipant? activeSpeaker;
+  WrappedMediaStream? localUserMediaStream;
+  WrappedMediaStream? localScreenshareStream;
+  String? localDesktopCapturerSourceId;
+  List<CallSession> callSessions = [];
+
+  CallParticipant? get localParticipant => voip.localParticipant;
+
+  /// userId:deviceId
+  List<CallParticipant> participants = [];
+  List<WrappedMediaStream> userMediaStreams = [];
+  List<WrappedMediaStream> screenshareStreams = [];
+  late String groupCallId;
+
+  GroupCallError? lastError;
+
+  Timer? activeSpeakerLoopTimeout;
+
+  Timer? resendMemberStateEventTimer;
+  Timer? memberLeaveEncKeyRotateDebounceTimer;
+
+  final CachedStreamController<GroupCallSession> onGroupCallFeedsChanged =
+      CachedStreamController();
+
+  final CachedStreamController<GroupCallState> onGroupCallState =
+      CachedStreamController();
+
+  final CachedStreamController<GroupCallEvent> onGroupCallEvent =
```

Comment:
How about you make this an actual event? As in have different classes, that have objects emitted, that describe the change like

```dart
class GroupCallMutedChanged {
  final bool muted;
}

class GroupCallStateChanged {
  final GroupCallState state;
}
```

etc?

Reply:
ehh, sounds a bit overkill imo, usually clients can just setState and flutter should handle everything if keys were used properly, maybe a future task? (not exactly new to this PR :D)

Reply:
Well, there is quite a bit of code, that listens to those events and then calls another function to check the proper state, which is error prone and confusing imo. But yeah, maybe that can be done later, although it would be another API break.

Reply:
moved to https://github.com/famedly/product-management/issues/1950 for now